config/v1/types_infrastructure.go

```diff
@@ -620,9 +620,36 @@ const (
 	AzureStackCloud AzureCloudEnvironment = "AzureStackCloud"
 )
 
+// GCPServiceEndpoint store the configuration of a custom url to
+// override existing defaults of GCP Services.
+type GCPServiceEndpoint struct {
+	// name is the name of the GCP service.
+	// The list of all the service names can be found at https://developers.google.com/apis-explorer
+	// This must be provided and cannot be empty.
+	//
+	// +kubebuilder:validation:Pattern=`^[a-z0-9-]+$`
+	// +required
+	Name string `json:"name"`
+
+	// url is fully qualified URI with scheme https, that overrides the default generated
+	// endpoint for a client.
+	// This must be provided and cannot be empty.
+	//
+	// +kubebuilder:validation:Pattern=`^https://`
+	// +required
+	URL string `json:"url"`
+}
+
 // GCPPlatformSpec holds the desired state of the Google Cloud Platform infrastructure provider.
 // This only includes fields that can be modified in the cluster.
-type GCPPlatformSpec struct{}
+type GCPPlatformSpec struct {
+	// serviceEndpoints list contains custom endpoints which will override default
+	// service endpoint of GCP Services.
+	// There must be only one ServiceEndpoint for a service.
```

Comment:
It sounds like this list must be unique based on the `GCPServiceEndpoint.Name`. I don't see any validation in place for this. I would recommend using CEL to validate that all items in this list are unique. When using CEL on a list, you'll need to add a max length for the list.

An example of uniqueness in CEL: https://github.com/openshift/api/blob/366ffb8c60da233a10af049e3869048531b7f6a1/example/v1/types_stable.go#L106-L107